tests/analyze_results.py

```diff
@@ -0,0 +1,129 @@
+import json
+import collections
+import functools
+import re
+import argparse
+
+
+@functools.total_ordering
+class RuleResult(object):
+    def __init__(self, result_dict):
+        self.scenario = (result_dict["rule_id"], result_dict["scenario_script"])
+        self.conditions = (
+            result_dict["backend"], result_dict["scanning_mode"],
+            result_dict["remediated_by"])
+        self.when = result_dict["run_timestamp"]
+        self.passed_stages = result_dict["passed_stages"]
+
+        self.success = result_dict.get("final_scan", False)
+        if not self.success:
+            self.success = (
+                "remediation" not in result_dict
+                and result_dict.get("initial_scan", False))
+
+    def relative_conditions_to(self, other):
+        if self.conditions == other.conditions:
+            return self.when, other.when
+        else:
+            return self.conditions, other.conditions
+
+    def __eq__(self, other):
+        return self.success == other.success and self.passed_stages == self.passed_stages
+
+    def __lt__(self, other):
+        return self.passed_stages > other.passed_stages
+
+
+class Difference(object):
+    def __init__(self, what_failed, why_failed, what_not_failed):
+        self.what_failed = what_failed
+        self.why_failed = why_failed
+        self.what_not_failed = what_not_failed
+
+    def __str__(self):
+        ret = ("{failed_stage} failed on {failed_config} and not on {good_config}"
+               .format(
+                   failed_stage=self.why_failed, failed_config=self.what_failed,
+                   good_config=self.what_not_failed))
+        return ret
+
+
+def aggregate_results_by_scenarios(rule_results):
+    aggregated = collections.defaultdict(list)
+    for result in rule_results:
+        aggregated[result.scenario].append(result)
+    return aggregated
+
+
+def analyze_differences(rules):
+    rules = sorted(rules)
```

Comment:
this is a bit cryptic and unexpected, we sort a list of rules and then only compare the lowest and highest. can you explain why? is rules expected to have more than 2 items?